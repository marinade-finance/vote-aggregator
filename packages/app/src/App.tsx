import {useCallback, useMemo} from 'react';
import {QueryClient, QueryClientProvider} from '@tanstack/react-query';
import {CssBaseline} from '@mui/material';
import {
  Adapter,
  WalletAdapterNetwork,
  WalletError,
} from '@solana/wallet-adapter-base';
import {clusterApiUrl} from '@solana/web3.js';
import {useSnackbar} from 'notistack';
import {ConnectionProvider, WalletProvider} from '@solana/wallet-adapter-react';
import {WalletDialogProvider} from '@solana/wallet-adapter-material-ui';
import Header from './components/Header';
import ClanList from './components/ClanList';

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      staleTime: 1000 * 60,
    },
  },
});

function App() {
  // The network can be set to 'devnet', 'testnet', or 'mainnet-beta'.
  const network = WalletAdapterNetwork.Devnet;

  // You can also provide a custom RPC endpoint.
  const endpoint = useMemo(() => clusterApiUrl(network), [network]);

  const {enqueueSnackbar} = useSnackbar();
  const onError = useCallback(
    (error: WalletError, adapter?: Adapter) => {
      enqueueSnackbar(
        error.message ? `${error.name}: ${error.message}` : error.name,
        {variant: 'error'}
      );
      console.error(error, adapter);
    },
    [enqueueSnackbar]
  );

  return (
    <QueryClientProvider client={queryClient}>
      <CssBaseline />
      <ConnectionProvider endpoint={endpoint}>
        <WalletProvider wallets={[]} onError={onError} autoConnect>
          <WalletDialogProvider>
            <Header />
            <ClanList sx={{width: '100%', ml: 10}} />
          </WalletDialogProvider>
        </WalletProvider>
      </ConnectionProvider>
    </QueryClientProvider>
  );
}

export default App;
