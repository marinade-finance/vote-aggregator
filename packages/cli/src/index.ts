import {cli} from './cli';

cli().parseAsync(process.argv);
