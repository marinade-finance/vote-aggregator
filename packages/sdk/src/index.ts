export * from './sdk';
export * from './root';
export * from './member';
export * from './clan';
