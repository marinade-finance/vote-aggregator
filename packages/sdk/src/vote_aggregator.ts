../../../target/types/vote_aggregator.ts