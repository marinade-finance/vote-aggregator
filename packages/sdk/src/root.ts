import {
  GetProgramAccountsFilter,
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from '@solana/web3.js';
import {RealmSide, VoteAggregatorSdk} from './sdk';
import {
  GoverningTokenConfig,
  MintMaxVoteWeightSource,
  createSetRealmConfig,
  getRealm,
  getRealmConfig,
  getRealmConfigAddress,
} from '@solana/spl-governance';
import {
  AnchorProvider,
  IdlAccounts,
  ProgramAccount,
} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {splGovernanceProgram} from '@coral-xyz/spl-governance';
import BN from 'bn.js';

export type RootAccount = IdlAccounts<VoteAggregator>['root'];
export type MaxVoterWeightAccount =
  IdlAccounts<VoteAggregator>['maxVoterWeightRecord'];

export class RootSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  rootAddress({
    realmAddress,
    governingTokenMint,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('root', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
      ],
      this.sdk.programId
    );
  }

  maxVoterWieghtAddress({
    rootAddress,
  }: {
    rootAddress: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('max-voter-weight', 'utf-8'), rootAddress.toBuffer()],
      this.sdk.programId
    );
  }

  lockAuthority({rootAddress}: {rootAddress: PublicKey}): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('lock-authority', 'utf-8'), rootAddress.toBuffer()],
      this.sdk.programId
    );
  }

  fetchRoot(rootAddress: PublicKey): Promise<RootAccount> {
    return this.sdk.program.account.root.fetch(rootAddress);
  }

  fetchRoots({
    governanceProgram,
    realm,
    governingTokenMint,
    votingWeightPlugin,
  }: {
    governanceProgram?: PublicKey;
    realm?: PublicKey;
    governingTokenMint?: PublicKey;
    votingWeightPlugin?: PublicKey;
  }): Promise<ProgramAccount<RootAccount>[]> {
    let filter: GetProgramAccountsFilter[] | undefined;

    if (governanceProgram) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 8,
          bytes: governanceProgram.toBase58(),
        },
      });
    }

    if (realm) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 40,
          bytes: realm.toBase58(),
        },
      });
    }

    if (governingTokenMint) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 72,
          bytes: governingTokenMint.toBase58(),
        },
      });
    }

    if (votingWeightPlugin) {
      if (!filter) {
        filter = [];
      }
      filter.push({
        memcmp: {
          offset: 104,
          bytes: votingWeightPlugin.toBase58(),
        },
      });
    }

    return this.sdk.program.account.root.all(filter);
  }

  fetchMaxVoterWeight({
    rootAddress,
    maxVoterWeightAddress,
  }: {
    rootAddress?: PublicKey;
    maxVoterWeightAddress?: PublicKey;
  }): Promise<MaxVoterWeightAccount> {
    if (!maxVoterWeightAddress) {
      if (!rootAddress) {
        throw new Error('rootAddress is required');
      }
      maxVoterWeightAddress = this.maxVoterWieghtAddress({rootAddress})[0];
    }
    return this.sdk.program.account.maxVoterWeightRecord.fetch(
      maxVoterWeightAddress
    );
  }

  async createRootInstructions({
    splGovernanceId,
    realmAddress,
    realmData,
    realmConfigData,
    side,
    maxProposalLifetime = new BN(0),
    payer,
  }: {
    splGovernanceId?: PublicKey;
    realmAddress: PublicKey;
    realmData?: {
      communityMint: PublicKey;
      config: {
        councilMint: PublicKey | undefined;
        communityMintMaxVoteWeightSource: MintMaxVoteWeightSource;
        minCommunityTokensToCreateGovernance: BN;
      };
      authority: PublicKey | undefined;
    };
    realmConfigData?: {
      communityTokenConfig: GoverningTokenConfig;
      councilTokenConfig: GoverningTokenConfig;
    };
    side: RealmSide;
    maxProposalLifetime?: BN;
    payer: PublicKey;
  }): Promise<TransactionInstruction[]> {
    if (!realmData) {
      const {account, owner} = await getRealm(
        this.sdk.connection,
        realmAddress
      );
      if (splGovernanceId && !owner.equals(splGovernanceId)) {
        throw new Error(
          `Realm ${realmAddress} is not owned by ${splGovernanceId}`
        );
      }
      realmData = account;
      splGovernanceId = owner;
    }
    if (!splGovernanceId) {
      throw new Error('splGovernanceId is required');
    }
    if (!realmData.authority) {
      throw new Error(`Realm ${realmAddress} does not have an authority`);
    }

    if (!realmConfigData) {
      const {account} = await getRealmConfig(
        this.sdk.connection,
        await getRealmConfigAddress(splGovernanceId, realmAddress)
      );
      realmConfigData = account;
    }

    const governingTokenMint =
      side === 'community'
        ? realmData.communityMint
        : realmData.config.councilMint!;
    if (!governingTokenMint) {
      throw new Error(`Realm ${realmAddress} does not have a ${side} mint`);
    }

    const [rootAddress] = this.rootAddress({realmAddress, governingTokenMint});
    const [maxVwr] = this.maxVoterWieghtAddress({rootAddress});

    const createRootIx = await this.sdk.program.methods
      .createRoot(maxProposalLifetime)
      .accountsStrict({
        root: rootAddress,
        realm: realmAddress,
        realmConfig: await getRealmConfigAddress(splGovernanceId, realmAddress),
        governingTokenMint,
        realmAuthority: realmData.authority,
        maxVwr,
        payer,
        governanceProgram: splGovernanceId,
        systemProgram: SystemProgram.programId,
        voteAggregatorProgram: this.sdk.program.programId,
      })
      .instruction();

    const communityTokenConfigArgs = {
      voterWeightAddin:
        side === 'community'
          ? this.sdk.program.programId
          : realmConfigData.communityTokenConfig.voterWeightAddin,
      maxVoterWeightAddin:
        realmConfigData.communityTokenConfig.maxVoterWeightAddin,
      tokenType: realmConfigData.communityTokenConfig.tokenType,
      useVoterWeightAddin:
        side === 'community' ||
        Boolean(realmConfigData.communityTokenConfig.voterWeightAddin),
      useMaxVoterWeightAddin: Boolean(
        realmConfigData.communityTokenConfig.maxVoterWeightAddin
      ),
    };
    const councilTokenConfigArgs = realmData.config.councilMint
      ? {
          voterWeightAddin:
            side === 'council'
              ? this.sdk.program.programId
              : realmConfigData.councilTokenConfig.voterWeightAddin,
          maxVoterWeightAddin:
            realmConfigData.councilTokenConfig.maxVoterWeightAddin,
          tokenType: realmConfigData.councilTokenConfig.tokenType,
          useVoterWeightAddin:
            side === 'council' ||
            Boolean(realmConfigData.councilTokenConfig.voterWeightAddin),
          useMaxVoterWeightAddin: Boolean(
            realmConfigData.councilTokenConfig.maxVoterWeightAddin
          ),
        }
      : undefined;

    const setRealmConfigIx = await createSetRealmConfig(
      splGovernanceId,
      3,
      realmAddress,
      realmData.authority!,
      realmData.config.councilMint || undefined,
      realmData.config.communityMintMaxVoteWeightSource,
      realmData.config.minCommunityTokensToCreateGovernance,
      communityTokenConfigArgs,
      councilTokenConfigArgs,
      payer
    );

    // TODO: use official JS SDK
    const splGovernance = splGovernanceProgram({
      programId: splGovernanceId,
      provider: new AnchorProvider(
        this.sdk.connection,
        {
          signTransaction: t => Promise.resolve(t),
          signAllTransactions: ts => Promise.resolve(ts),
          publicKey: PublicKey.default,
        },
        {}
      ),
    });
    const [lockAuthority] = this.lockAuthority({rootAddress});
    const setLockAuthorityIx = await splGovernance.methods
      .setRealmConfigItem({
        tokenOwnerRecordLockAuthority: {
          action: {
            add: {},
          },
          governingTokenMint,
          authority: lockAuthority,
        },
      })
      .accountsStrict({
        realm: realmAddress,
        realmAuthority: realmData.authority!,
        payer,
        systemProgram: SystemProgram.programId,
        realmConfigAddress: await getRealmConfigAddress(
          splGovernanceId,
          realmAddress
        ),
      })
      .instruction();

    return [createRootIx, setRealmConfigIx, setLockAuthorityIx];
  }

  async setMaxProposalLifetimeInstruction({
    maxProposalLifetime,
    root,
    realm,
    realmAuthority,
  }: {
    maxProposalLifetime: BN;
    root: PublicKey;
    realm: PublicKey;
    realmAuthority: PublicKey;
  }) {
    return await this.sdk.program.methods
      .setMaxProposalLifetime(maxProposalLifetime)
      .accountsStrict({
        realm,
        realmAuthority,
        root,
      })
      .instruction();
  }

  async setVoterWeightResetInstruction({
    step,
    nextResetTime,
    root,
    realm,
    realmAuthority,
  }: {
    step: BN;
    nextResetTime: BN | null;
    root: PublicKey;
    realm: PublicKey;
    realmAuthority: PublicKey;
  }) {
    return await this.sdk.program.methods
      .setVoterWeightReset(step, nextResetTime)
      .accountsStrict({
        realm,
        realmAuthority,
        root,
      })
      .instruction();
  }
}
