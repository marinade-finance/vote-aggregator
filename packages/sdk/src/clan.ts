import {PublicKey, SystemProgram} from '@solana/web3.js';
import {VoteAggregatorSdk} from './sdk';
import {IdlAccounts} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {RootAccount} from './root';

export type ClanAccount = IdlAccounts<VoteAggregator>['clan'];
export type VoterWeightAccount =
  IdlAccounts<VoteAggregator>['voterWeightRecord'];

export class ClanSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  voterAuthority({clanAddress}: {clanAddress: PublicKey}): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('voter-authority', 'utf-8'), clanAddress.toBuffer()],
      this.sdk.programId
    );
  }

  tokenOwnerRecordAddress({
    realmAddress,
    governingTokenMint,
    clanAddress,
    voterAuthority,
    splGovernanceId,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
    clanAddress?: PublicKey;
    voterAuthority?: PublicKey;
    splGovernanceId: PublicKey;
  }): [PublicKey, number] {
    if (!voterAuthority) {
      if (!clanAddress) {
        throw new Error('clanAddress is required');
      }
      voterAuthority = this.voterAuthority({clanAddress})[0];
    }
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('governance', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
        voterAuthority.toBuffer(),
      ],
      splGovernanceId
    );
  }

  voterWeightAddress(clanAddress: PublicKey): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [Buffer.from('voter-weight', 'utf-8'), clanAddress.toBuffer()],
      this.sdk.programId
    );
  }

  fetchClan(clanAddress: PublicKey): Promise<ClanAccount> {
    return this.sdk.program.account.clan.fetch(clanAddress);
  }

  fetchVoterWeight({
    clanAddress,
    voterWeightAddress,
  }: {
    clanAddress?: PublicKey;
    voterWeightAddress?: PublicKey;
  }): Promise<VoterWeightAccount> {
    if (!voterWeightAddress) {
      if (!clanAddress) {
        throw new Error('clanAddress is required');
      }
      voterWeightAddress = this.voterWeightAddress(clanAddress)[0];
    }
    return this.sdk.program.account.voterWeightRecord.fetch(voterWeightAddress);
  }

  async createClanInstruction({
    rootAddress,
    root,
    clanAddress,
    owner,
    payer,
  }: {
    rootAddress: PublicKey;
    root: RootAccount;
    clanAddress: PublicKey;
    owner: PublicKey;
    payer: PublicKey;
  }) {
    const [voterAuthority] = this.voterAuthority({clanAddress});
    const [tokenOwnerRecord] = this.tokenOwnerRecordAddress({
      realmAddress: root.realm,
      governingTokenMint: root.governingTokenMint,
      clanAddress,
      splGovernanceId: root.governanceProgram,
    });
    const [voterWeightRecord] = this.voterWeightAddress(clanAddress);
    return await this.sdk.program.methods
      .createClan(owner)
      .accountsStrict({
        root: rootAddress,
        clan: clanAddress,
        realm: root.realm,
        governingTokenMint: root.governingTokenMint,
        payer,
        governanceProgram: root.governanceProgram,
        systemProgram: SystemProgram.programId,
        voterAuthority,
        tokenOwnerRecord,
        voterWeightRecord,
      })
      .instruction();
  }
}
