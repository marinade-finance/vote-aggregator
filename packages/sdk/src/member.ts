import {PublicKey, SystemProgram} from '@solana/web3.js';
import {VoteAggregatorSdk} from './sdk';
import {BN, IdlAccounts} from '@coral-xyz/anchor';
import {VoteAggregator} from './vote_aggregator';
import {RootAccount} from './root';
import {SYSTEM_PROGRAM_ID} from '@solana/spl-governance';
import {VoterWeightAccount} from './clan';

export type MemberAccount = IdlAccounts<VoteAggregator>['member'];

export class MemberSdk {
  constructor(public readonly sdk: VoteAggregatorSdk) {}

  memberAddress({
    rootAddress,
    owner,
  }: {
    rootAddress: PublicKey;
    owner: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('member', 'utf-8'),
        rootAddress.toBuffer(),
        owner.toBuffer(),
      ],
      this.sdk.programId
    );
  }

  tokenOwnerRecordAddress({
    realmAddress,
    governingTokenMint,
    owner,
    splGovernanceId,
  }: {
    realmAddress: PublicKey;
    governingTokenMint: PublicKey;
    owner: PublicKey;
    splGovernanceId: PublicKey;
  }): [PublicKey, number] {
    return PublicKey.findProgramAddressSync(
      [
        Buffer.from('governance', 'utf-8'),
        realmAddress.toBuffer(),
        governingTokenMint.toBuffer(),
        owner.toBuffer(),
      ],
      splGovernanceId
    );
  }

  fetchMember({
    memberAddress,
    owner,
    rootAddress,
  }: {
    memberAddress?: PublicKey;
    rootAddress?: PublicKey;
    owner?: PublicKey;
  }): Promise<MemberAccount> {
    if (!memberAddress) {
      if (!owner) {
        throw new Error('owner is required');
      }
      if (!rootAddress) {
        throw new Error('rootAddress is required');
      }
      memberAddress = this.memberAddress({rootAddress, owner})[0];
    }
    return this.sdk.program.account.member.fetch(memberAddress);
  }

  async findVoterWeightRecord({
    root,
    member,
  }: {
    root: RootAccount;
    member: MemberAccount;
  }) {
    const voterWeightAccounts = await this.sdk.connection.getProgramAccounts(
      root.votingWeightPlugin,
      {
        filters: [
          {
            memcmp: {
              offset: 0,
              bytes: '8riZd8mYDQk', // Discriminator
            },
          },
          {
            memcmp: {
              offset: 8,
              bytes: root.realm.toBase58(),
            },
          },
          {
            memcmp: {
              offset: 8 + 32,
              bytes: root.governingTokenMint.toBase58(),
            },
          },
          {
            memcmp: {
              offset: 8 + 32 + 32,
              bytes: member.owner.toBase58(),
            },
          },
        ],
      }
    );

    const voterWeightRecords = voterWeightAccounts.map(({account, pubkey}) => {
      const record = this.sdk.program.coder.accounts.decode<VoterWeightAccount>(
        'voterWeightRecord',
        account.data
      );
      return {
        record,
        pubkey,
      };
    });

    let maxExpiry: BN | null = null;
    let maxPower = new BN(0);
    let bestIndex = null;
    for (let i = 0; i < voterWeightRecords.length; i++) {
      const {record} = voterWeightRecords[i];
      if (record.weightAction || record.weightActionTarget) {
        continue;
      }
      if (
        (!record.voterWeightExpiry && maxExpiry) ||
        (maxExpiry && record.voterWeightExpiry?.gt(maxExpiry))
      ) {
        maxExpiry = record.voterWeightExpiry;
        maxPower = record.voterWeight;
        bestIndex = i;
      } else if (
        (record.voterWeightExpiry === null && maxExpiry === null) ||
        (maxExpiry && record.voterWeightExpiry?.eq(maxExpiry))
      ) {
        if (record.voterWeight.gt(maxPower)) {
          maxPower = record.voterWeight;
          bestIndex = i;
        }
      }
    }

    if (bestIndex === null) {
      throw new Error(
        `Can not find VWR for ${member.owner.toBase58()} in plugin ${root.votingWeightPlugin.toBase58()}`
      );
    }

    return voterWeightRecords[bestIndex];
  }

  async createMemberInstruction({
    rootAddress,
    root,
    owner,
    payer,
  }: {
    rootAddress: PublicKey;
    root: RootAccount;
    owner: PublicKey;
    payer: PublicKey;
  }) {
    const [tokenOwnerRecord] = this.tokenOwnerRecordAddress({
      realmAddress: root.realm,
      governingTokenMint: root.governingTokenMint,
      owner,
      splGovernanceId: root.governanceProgram,
    });
    const [member] = this.memberAddress({rootAddress, owner});
    return await this.sdk.program.methods
      .createMember()
      .accountsStrict({
        root: rootAddress,
        member,
        payer,
        systemProgram: SYSTEM_PROGRAM_ID,
        tokenOwnerRecord,
        owner,
      })
      .instruction();
  }

  async joinClanInstruction({
    root,
    member,
    clanAddress,
    memberVoterWeightAddress,
  }: {
    root: RootAccount;
    member: MemberAccount;
    clanAddress: PublicKey;
    memberVoterWeightAddress: PublicKey;
  }) {
    const [memberAddress] = this.memberAddress({
      rootAddress: member.root,
      owner: member.owner,
    });
    return await this.sdk.program.methods
      .joinClan()
      .accountsStrict({
        root: member.root,
        member: memberAddress,
        clan: clanAddress,
        memberAuthority: member.owner, // TODO delegate
        clanVoterWeightRecord: this.sdk.clan.voterWeightAddress(clanAddress)[0],
        memberTokenOwnerRecord: this.tokenOwnerRecordAddress({
          realmAddress: root.realm,
          governingTokenMint: root.governingTokenMint,
          owner: member.owner,
          splGovernanceId: root.governanceProgram,
        })[0],
        memberVoterWeightRecord: memberVoterWeightAddress,
        maxVoterWeightRecord: this.sdk.root.maxVoterWieghtAddress({
          rootAddress: member.root,
        })[0],
      })
      .instruction();
  }
}
