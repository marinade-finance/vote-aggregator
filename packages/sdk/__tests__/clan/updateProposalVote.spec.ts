import {describe, it, expect} from 'bun:test';
import {VoteAggregatorSdk} from '../../src';
import {
  UpdateProposalVoteTestData,
  RealmTester,
  updateProposalVoteTestData,
} from 'vote-aggregator-tests';
import {ClanTester, RootTester} from 'vote-aggregator-tests';
import {GovernanceTester} from 'vote-aggregator-tests/src/SplGovernance/governance';
import {ProposalTester} from 'vote-aggregator-tests/src/SplGovernance';
import {PublicKey} from '@solana/web3.js';

describe('update_proposal_vote instruction', () => {
  it.each(updateProposalVoteTestData.filter(({error}) => !error))(
    'Works',
    async ({
      realm,
      root,
      clan,
      governance,
      proposal,
    }: UpdateProposalVoteTestData) => {
      const realmTester = new RealmTester(realm);
      const rootTester = new RootTester({
        ...root,
        realm: realmTester,
      });
      const clanTester = new ClanTester({...clan, root: rootTester});
      const governanceTester = new GovernanceTester({
        ...governance,
        realm: realmTester,
      });
      const proposalTester = new ProposalTester({
        ...proposal,
        governance: governanceTester,
        clan: proposal.owner ? undefined : clanTester,
      });

      const sdk = new VoteAggregatorSdk();
      expect(
        sdk.clan.udpateProposalVote({
          rootAddress: rootTester.rootAddress[0],
          rootData: rootTester.root,
          clanAddress: clan.address,
          governance: governanceTester.governanceAddress,
          proposal: proposalTester.proposalAddress,
          proposalOwnerRecord: proposalTester.proposal.tokenOwnerRecord,
          payer: PublicKey.default,
        })
      ).resolves.toMatchSnapshot();
    }
  );
});
