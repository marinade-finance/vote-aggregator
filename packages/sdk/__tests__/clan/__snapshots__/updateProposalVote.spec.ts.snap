// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`update_proposal_vote instruction Works 1`] = `
TransactionInstruction {
  "data": {
    "data": [
      22,
      86,
      173,
      132,
      228,
      210,
      48,
      151,
    ],
    "type": "Buffer",
  },
  "keys": [
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(5nYvDD2VT4Zx1kxNohfATAAMMxpajEQkaD4vyHrNyGJR) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            61969086,
            56097209,
            38522167,
            22027141,
            44359793,
            59094752,
            10403524,
            46903005,
            19360877,
            1164926,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(5HNh2T1Xv3RNwQuSoUXhdy6UZPr9VYZBHQuWzrGZSUqx) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            18123755,
            5757342,
            5388769,
            50592869,
            3954940,
            31854970,
            57136948,
            48058040,
            13781455,
            1042468,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(GCfsGVkUtbJTXC1uU6qNTktghf1ND4Mn1gqWYHrKU7c3) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            14834240,
            58035910,
            49412216,
            53469220,
            60481551,
            56324336,
            54517476,
            44771747,
            62212466,
            3700520,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(2xmdXeiV9azNt2aW9FTPAuwnrJ2Lb4gZqtHsN1A8EP6v) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            55693063,
            54538942,
            8604916,
            4398723,
            10744168,
            40695456,
            1353329,
            9859640,
            3230881,
            477510,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(4uynsQ25jPGzpK7ZwFYB5R7Qvy6reXe3pRBpws3sFE8G) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            26693001,
            49501911,
            41732767,
            19421841,
            66678386,
            44254438,
            54674187,
            19967447,
            33123642,
            952670,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(2TrqzfsojMCXBpR2RoRGf6vJ39hP1fcYZ3qvzTSU5tAc) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            10792009,
            28379722,
            9404275,
            8819746,
            43112967,
            1993951,
            39191264,
            51473836,
            29861645,
            356169,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(YmoYimqAG2EVcn5RdW9od3jXNNEGRLvg65yCGkuqWMY) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            46874523,
            64651139,
            15443642,
            47894726,
            13477292,
            29938569,
            43087945,
            43514584,
            30995171,
            133354,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(BFCAhwpbrNZfPtNzGtf14ckeXcDRYCzNC2zBKX8LLCh4) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            59503271,
            34072060,
            48323839,
            19453488,
            31292829,
            42872141,
            18572706,
            10235021,
            10937907,
            2493928,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(J7HEwThZ62S789TdBCATixaNMvovjXnN88qjH4moxSFN) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            53363581,
            60481416,
            10601011,
            11957757,
            13472368,
            54810654,
            2037198,
            63010609,
            42837245,
            4164769,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(CHUSDriSHbrVn8JeAveYC4dgdsxvmtpM4i8rykQkbTvL) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            3635677,
            3059973,
            15308077,
            6724564,
            22941632,
            56071264,
            58417319,
            19405728,
            44887460,
            2746935,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(5mmWyxTexBSM4XLPfSmiNoG5e8rZpGcZoBfyxKE4W9BD) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            56680520,
            30718302,
            58522471,
            11044063,
            38497358,
            57996517,
            35585688,
            39289883,
            59095189,
            1161640,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(VoTaGDreyne7jk59uwbgRRbaAzxvNbyNipaJMrRXhjT) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            8873870,
            60943108,
            23935113,
            33099803,
            59555925,
            65803128,
            11262563,
            66819466,
            50754419,
            120882,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": true,
      "pubkey": PublicKey(79RLV4UPM7BWR1WwJmhsCvvM3SVYCJAwWWevATS56X6o) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            4461360,
            58977221,
            15565329,
            2029576,
            49443948,
            63907247,
            6603228,
            40628402,
            15039176,
            1495954,
            0,
          ],
        },
      },
    },
    {
      "isSigner": true,
      "isWritable": true,
      "pubkey": PublicKey(11111111111111111111111111111111) PublicKey {
        "_bn": BN {
          "length": 1,
          "negative": 0,
          "red": null,
          "words": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(11111111111111111111111111111111) PublicKey {
        "_bn": BN {
          "length": 1,
          "negative": 0,
          "red": null,
          "words": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ],
        },
      },
    },
    {
      "isSigner": false,
      "isWritable": false,
      "pubkey": PublicKey(3xNbuehooXRpHEbntV4BCwaaMdLHrGcfvjdopTA6idRq) PublicKey {
        "_bn": BN {
          "length": 10,
          "negative": 0,
          "red": null,
          "words": [
            10823944,
            60529926,
            696011,
            7145378,
            4135132,
            53082547,
            30530522,
            52587429,
            39837786,
            719279,
            0,
          ],
        },
      },
    },
  ],
  "programId": PublicKey(VoTaGDreyne7jk59uwbgRRbaAzxvNbyNipaJMrRXhjT) PublicKey {
    "_bn": BN {
      "length": 10,
      "negative": 0,
      "red": null,
      "words": [
        8873870,
        60943108,
        23935113,
        33099803,
        59555925,
        65803128,
        11262563,
        66819466,
        50754419,
        120882,
        0,
      ],
    },
  },
}
`;
